libsolidity/analysis/TypeChecker.cpp

```diff
@@ -810,12 +811,20 @@ bool TypeChecker::visit(TupleExpression const& _tuple)
 			{
 				components[i]->accept(*this);
 				types.push_back(type(*components[i]));
+				if (i == 0 && isArray)
+					internalArrayType = types[i]->mobileType();
+				else if (isArray && internalArrayType && types[i]->mobileType())
+					internalArrayType = Type::commonType(internalArrayType, types[i]->mobileType());
 			}
 			else
 				types.push_back(TypePointer());
 		}
-		if (components.size() == 1)
+		if (components.size() == 1 && !isArray)
```

Comment:
This would perhaps be easier to read if you first check for `isArray` and then to the subsequent checks.